x/programs/rust/wasmlanche-sdk/src/state.rs

```diff
@@ -154,13 +191,24 @@ impl<'a, K: Key> State<'a, K> {
             value: Vec<u8>,
         }
 
-        let mut cache = self.cache.borrow_mut();
-
-        let args: Vec<_> = cache
+        let (mut puts, mut deletes) = (Vec::new(), Vec::new());
+        self.cache
+            .borrow_mut()
             .drain()
-            .map(|(key, value)| PutArgs { key, value })
-            .collect();
-        let serialized_args = borsh::to_vec(&args).expect("failed to serialize");
-        unsafe { put_many_bytes(serialized_args.as_ptr(), serialized_args.len()) };
+            .for_each(|(key, value)| match value {
+                CachedData::Data(value) => puts.push(PutArgs { key, value }),
+                CachedData::ToDelete => deletes.push(key),
```

Comment:
Yeah, this tells me that a single API here would be better. I can't remember where I said it (could have been in this review or elsewhere), but I think that putting an empty `Vec<u8>` isn't valid. So you can actually eliminate all invalid state (minimizing the runtime checks) by making empty `Vec` puts the equivalent of `delete`. 

@dboehm-avalabs, thoughts?